Documentation/Sanoid.5
.\" Automatically generated by Pandoc 2.17.1.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "Sanoid.json" "5" "June 2023" "Sanoid.net 1.0.0-Beta1 Configuration" ""
.hy
.SH Name
.PP
Sanoid.json - Configuration files for Sanoid.net
.SH Synopsis
.PP
\f[B]/usr/local/share/Sanoid.net/Sanoid.json\f[R] (Sanoid.json)
.PD 0
.P
.PD
\f[B]/etc/sanoid/Sanoid.local.json\f[R] (Sanoid.local.json)
.SH Description
.PP
These files contain general configuration and templates for Sanoid.net.
.PP
All configuration files are formatted as plain-text JSON files, in UTF-8
encoding.
.PP
Note: Command line arguments which affect settings in any configuration
file have highest precedence and will override their matching settings,
at run-time.
.PD 0
.P
.PD
See Sanoid.net(8) for details on command line arguments.
.SS Configuration Files
.PP
\f[B]/usr/local/share/Sanoid.net/Sanoid.json\f[R] is the base
configuration file and is not intended, recommended, or supported to be
changed by the end user.
.PD 0
.P
.PD
This file is installed with read-only permissions, by default, to help
avoid unintended modifications.
.PP
\f[B]/etc/sanoid/Sanoid.local.json\f[R] is the local configuration file.
.PD 0
.P
.PD
All settings specified in this file extend the configuration contained
in Sanoid.json.
Any leaf-level settings in Sanoid.local.json that have the same
fully-qualified JSON path as leaf-level settings defined in Sanoid.json
will override the matching settings in Sanoid.json, at run-time.
.SS Configuration File Schemas
.PP
The schemas for Sanoid.json and Sanoid.local.json are identical, except
for required elements.
Both can be found in /usr/local/share/Sanoid.net as Sanoid.schema.json
and Sanoid.local.schema.json.
.PD 0
.P
.PD
The schema documents fully describe the layout and legal format for
these configuration files, and are to be considered authoritative, if
any discrepancies exist between this document and those files.
.PP
Changes to the schema documents are NOT supported and are likely to
cause run-time errors in Sanoid.net.
DO NOT MODIFY THE SCHEMA DOCUMENTS.
.SH Settings
.PP
Settings are described below using their JSON path as identifiers, and a
description including type, default value, and what each settings does.
.SS General Settings
.PP
These are general settings that affect the behavior of Sanoid.net.
.PD 0
.P
.PD
They can be overridden by command-line arguments, as noted.
.TP
\f[B]/DryRun\f[R] (boolean - \[lq]true\[rq] or \[lq]false\[rq])
This is a global setting that controls whether Sanoid.net will perform a
dry run, when invoked.
If this setting is \[lq]true,\[rq] Sanoid.net will perform a dry run and
make no changes to ZFS.
The command-line option \f[B]--dry-run\f[R] overrides this setting.
.TP
\f[B]/TakeSnapshots\f[R] (boolean - \[lq]true\[rq] or \[lq]false\[rq])
This is a global setting that controls whether Sanoid.net will take new
snapshots.
If this setting is true, Sanoid.net will take snapshots according to the
properties set on each ZFS dataset.
If this setting is false, new snapshot processing is skipped and no new
snapshots will be taken, regardless of the sanoid.net:takesnapshots
property set on each ZFS dataset.
The command-line options \f[B]--take-snapshots\f[R] and
\f[B]--no-take-snapshots\f[R] override this setting.
.TP
\f[B]/PruneSnapshots\f[R] (boolean - \[lq]true\[rq] or \[lq]false\[rq])
This is a global setting that controls whether Sanoid.net will prune
expired snapshots.
If this setting is true, Sanoid.net will prune eligible snapshots
according to the properties set on each snapshot or inherited from its
parent dataset.
If this setting is false, existing snapshots will not be pruned,
regardless of the sanoid.net:prunesnapshots property setting on each ZFS
dataset, the retention properties on each ZFS dataset or snapshot, or
the age of the snapshot.
The command-line options \f[B]--prune-snapshots\f[R] and
\f[B]--no-prune-snapshots\f[R] override this setting.
.SS Templates
.PP
Templates, which provide common timing and formatting settings for
snapshots, are defined under the \f[B]/Templates\f[R] node, which is a
dcitionary of key:value pairs, where the key is the template\[cq]s name,
as a quoted string, and the value is the template itself, as a JSON
object, with a schema defined in Sanoid.template.schema.json.
.PP
For a template to be valid for use, it must be fully defined, with all
options explicitly set.
As with other configuration elements, it is possible to configure
templates manually, in the json files, but it is \f[B]strongly\f[R]
recommended that you use the configuration console to do so, instead, as
it will ensure your templates are fully and properly defined, and will
prevent you from deleting templates that are in use by any datasets in
ZFS.
.PP
A template is a JSON object consisting of two sections:
\[lq]Formatting\[rq] and \[lq]SnapshotTiming\[rq]
.SS Template Formatting Section
.PP
The \f[B]Formatting\f[R] section of a template defines elements used for
naming of new snapshots that Sanoid.net creates.
The default Formatting options provided in the \[lq]default\[rq]
template, on a new install, will result in snapshot names that follow
the same format that PERL sanoid creates.
For example, a daily snapshot taken on 2023-06-22 at 14:00:00 (local
time) will be named
\[lq]path/to/dataset\[at]\f[B]autosnap_2023-06-22_14:00:00_daily\f[R]\[rq].
.PP
These settings do not affect operation of Sanoid.net in any way
\f[I]other\f[R] than in how it will name snapshots created on datasets
with that template applied.
.PP
Pruning of existing snapshots is in no way affected by these settings,
as Sanoid.net keeps metadata defining what a snapshot \f[I]is\f[R] in
ZFS properties, which are not affected by these settings.
.PP
If you use a heterogeneous setup combining both PERL sanoid and
Sanoid.net, you MUST use Formatting settings that correspond identically
to PERL sanoid\[cq]s configuration, or else PERL sanoid will
misinterpret or skip them entirely, as PERL sanoid decides what a
snapshot is and what to do with it based solely off of its name.
.PP
Note that all settings must conform to ZFS identifier rules, as they
apply to naming of snapshots.
Thus, only 7-bit ASCII characters in restricted ranges are supported.
The configuration console helps enforce these restrictions.
See zfs(8) for specifics.
.TP
\f[B]/Templates/templateName/Formatting/ComponentSeparator\f[R]
This is a single character used to separate each component of a
snapshot\[cq]s name.
By default, it is the underscore (_) character.
.TP
\f[B]/Templates/templateName/Formatting/Prefix\f[R]
The prefix is the first component of a snapshot name, appearing
immediately after the \[at] symbol.
It should be a 1-12 character alphanumeric string and is case-sensitive.
By default, it is the string \[lq]autosnap\[rq]
.TP
\f[B]/Templates/templateName/Formatting/TimestampFormatString\f[R]
The Timestamp Format String is a special format string that determines
how the timestamp is formatted in the name of a snapshot.
Sanoid.net uses the .net DateTimeOffset type, internally, for
representation of time, which contains the full date, time, and timezone
offset (in local time, by default) that the operation was executed at.
This string must be escaped, following both JSON and C# rules (which are
mostly the same).
The default string is \[lq]yyyy-MM-dd_HH\[rs]:mm\[rs]:ss\[rq], which
results in output as shown in the example given above.
Other format strings are allowed, but they must adhere to ZFS snapshot
identifier requirements and must be valid .net DateTimeOffset format
specifier strings.
Documentation of valid format strings can be found at
https://learn.microsoft.com/en-us/dotnet/standard/base-types/standard-date-and-time-format-strings
.TP
\f[B]/Templates/templateName/Formatting/FrequentSuffix\f[R]
The Frequent Suffix is the string used as the final portion of a
snapshot name, when that snapshot\[cq]s period is \[lq]frequent.\[rq] It
should be a 1-12 character alphanumeric string and is case-sensitive.
By default, it is the string \[lq]frequently\[rq]
.TP
\f[B]/Templates/templateName/Formatting/HourlySuffix\f[R]
The Hourly Suffix is the string used as the final portion of a snapshot
name, when that snapshot\[cq]s period is \[lq]hourly.\[rq] It should be
a 1-12 character alphanumeric string and is case-sensitive.
By default, it is the string \[lq]hourly\[rq]
.TP
\f[B]/Templates/templateName/Formatting/DailySuffix\f[R]
The Daily Suffix is the string used as the final portion of a snapshot
name, when that snapshot\[cq]s period is \[lq]daily.\[rq] It should be a
1-12 character alphanumeric string and is case-sensitive.
By default, it is the string \[lq]daily\[rq]
.TP
\f[B]/Templates/templateName/Formatting/WeeklySuffix\f[R]
The Weekly Suffix is the string used as the final portion of a snapshot
name, when that snapshot\[cq]s period is \[lq]weekly.\[rq] It should be
a 1-12 character alphanumeric string and is case-sensitive.
By default, it is the string \[lq]weekly\[rq]
.TP
\f[B]/Templates/templateName/Formatting/MonthlySuffix\f[R]
The Monthly Suffix is the string used as the final portion of a snapshot
name, when that snapshot\[cq]s period is \[lq]monthly.\[rq] It should be
a 1-12 character alphanumeric string and is case-sensitive.
By default, it is the string \[lq]monthly\[rq]
.TP
\f[B]/Templates/templateName/Formatting/YearlySuffix\f[R]
The Yearly Suffix is the string used as the final portion of a snapshot
name, when that snapshot\[cq]s period is \[lq]yearly.\[rq] It should be
a 1-12 character alphanumeric string and is case-sensitive.
By default, it is the string \[lq]yearly\[rq]
.SS Template Snapshot Timing Section
.PP
The \f[B]SnapshotTiming\f[R] section of a template allows you to
fine-tune the times at which Sanoid.net will take specific types of
snapshots, and when existing snapshots will be considered eligible for
pruning.
.PP
Default settings correspond to the same behavior that PERL sanoid
exhibits.
.TP
\f[B]/Templates/templateName/SnapshotTiming/UseLocalTime\f[R]
This setting is currently not used by Sanoid.net and is reserved for
future changes.
When this setting is implemented, it will control whether Sanoid.net
uses local system time (true) or UTC (false) for snapshot timestamps and
all associated processing, such as calculating eligibility for pruning.
The current behavior is the same as a setting of \[lq]true\[rq]
.TP
\f[B]/Templates/templateName/SnapshotTiming/FrequentPeriod\f[R]
This setting is a period, in minutes, for frequent snapshots to be
processed.
This value should be a whole-number factor of 60, such as 5, 10, 15, or
20.
While other values may work, they are not recommended nor are they
supported.
The value must be an un-quoted integer from 1 to 59.
.TP
\f[B]/Templates/templateName/SnapshotTiming/HourlyMinute\f[R]
This setting is the minute of the hour on which Sanoid.net will take
hourly snapshots.
This value must be an un-quoted whole number from 0 to 59.
The default value is 0, meaning that snapshots will be taken at the top
of the following hour.
For example, the hourly snapshot for the hour period from 11:00 to 11:59
will be taken at 12:00.
.TP
\f[B]/Templates/templateName/SnapshotTiming/DailyTime\f[R]
This setting is a time string, in HH:mm:ss format (0-fill required for
all components), and is the time of day that daily snapshots will be
taken.
Time strings can be any valid time of day from 00:00:00 to 23:59:59 and
may optionally include fractional seconds, up to 7 decimal places,
though the accuracy of actual snapshot timing will depend on the
precision of your system\[cq]s clock and the precision of the mechanism
used to invoke Sanoid.net.
For example, a time string of \[lq]12:34:56.789\[rq] is perfectly legal.
This will be more reliable once Sanoid.net has the ability to run as a
daemon.
.TP
\f[B]/Templates/templateName/SnapshotTiming/WeeklyDay\f[R]
This setting is a number, from 0 to 6, specifying the day of the week on
which weekly snapshots will be taken.
Note that Sanoid.net attempts to be culture-aware, and the meaning of
the number may depend on your system locale\[cq]s definition of a week
and which day is the start of that week.
In the invariant or US culture, 0 is Sunday.
The default setting is 1, which, in the invariant or US culture is
Monday, which is the default behavior of PERL sanoid, as well.
The configuraiton colsole will present this setting to you as the name
of the days of the week, in the system locale\[cq]s language and
calendar.
.TP
\f[B]/Templates/templateName/SnapshotTiming/WeeklyTime\f[R]
As with DailyTime, this setting is the time of day at which weekly
snapshots will be taken.
The rules, restrictions, and default value are the same as for
DailyTime.
See DailyTime for details.
.TP
\f[B]/Templates/templateName/SnapshotTiming/MonthlyDay\f[R]
This setting is the day of the month on which monthly snapshots will be
taken.
It is a number from 1 to the system locale\[cq]s maximum day number for
a month.
If the value is set higher than a given month\[cq]s last day, snapshots
will be taken on the last day of the month (so a setting of 31 will
always take a snapshot on the last day of the month, on the Gregorian
calendar).
The default value is 1, meaning the first day of every month.
.TP
\f[B]/Templates/templateName/SnapshotTiming/MonthlyTime\f[R]
As with DailyTime, this setting is the time of day at which monthly
snapshots will be taken.
The rules, restrictions, and default value are the same as for
DailyTime.
See DailyTime for details.
.TP
\f[B]/Templates/templateName/SnapshotTiming/YearlyMonth\f[R]
This setting is the number of the month of the year in which yearly
snapshots will be taken.
It is a whole-number value from 1 to the number corresponding to the
last month of the year, in the system locale\[cq]s calendar (12, for
Gregorian).
The configuration console will present this setting to you as the names
of the months of the year, in the system locale\[cq]s language and
calendar.
.TP
\f[B]/Templates/templateName/SnapshotTiming/YearlyDay\f[R]
As with MonthlyDay, this settings is the day of the month specified in
YearlyMonth on which yearly snapshots will be taken.
The rules, restrictions, and default are the same as for MonthlyDay.
See MonthlyDay for details.
.TP
\f[B]/Templates/templateName/SnapshotTiming/YearlyTime\f[R]
As with DailyTime, this setting is the time of day at which yearly
snapshots will be taken, on the YearlyMonth and YearlyDay specified.
The rules, restrictions, and default value are the same as for
DailyTime.
See DailyTime for details.
.SH Examples
.PP
These are example valid configurations.
.SS Default Base Configuration (/usr/local/share/Sanoid.net/Sanoid.json)
.PP
This is the default configuration shipped in Sanoid.json, which is the
base configuration file Sanoid.net builds the rest of its configuration
from and is not intended to be modified by the user.
Elements beginning with a $ symbol are metadata and do not affect
operation of Sanoid.net itself (though they should not be modified by
the user, as that may affect schema validation).
.PP
This configuration results in a default invocation of Sanoid.net
performing no actions that change ZFS (snapshots and pruning are
disabled).
Naming and timing settings for the default template included in this
configuration are the same as PERL sanoid\[cq]s defaults.
.IP
.nf
\f[C]
{
  \[dq]$schema\[dq]: \[dq]Sanoid.schema.json\[dq],
  \[dq]$id\[dq]: \[dq]Sanoid.json\[dq],
  \[dq]$comments\[dq]: \[dq]Default settings for sanoid.net. It is not recommended to modify this file. Customized settings should be specified in /etc/sanoid/Sanoid.local.json\[dq],
  \[dq]TakeSnapshots\[dq]: false,
  \[dq]PruneSnapshots\[dq]: false,
  \[dq]CacheDirectory\[dq]: \[dq]/var/cache/sanoid\[dq],
  \[dq]ZfsPath\[dq]: \[dq]/usr/local/sbin/zfs\[dq],
  \[dq]ZpoolPath\[dq]: \[dq]/usr/local/sbin/zpool\[dq],
  \[dq]DryRun\[dq]: false,
  \[dq]Monitoring\[dq]: {
    \[dq]Nagios\[dq]: {
      \[dq]$comments\[dq]: \[dq]Nagios-specific monitoring options\[dq],
      \[dq]MonitorType\[dq]: \[dq]Nagios\[dq],
      \[dq]Capacity\[dq]: false,
      \[dq]Health\[dq]: false,
      \[dq]Snapshots\[dq]: false
    }
  },
  \[dq]Templates\[dq]: {
    \[dq]default\[dq]: {
      \[dq]Formatting\[dq]: {
        \[dq]ComponentSeparator\[dq]: \[dq]_\[dq],
        \[dq]Prefix\[dq]: \[dq]autosnap\[dq],
        \[dq]TimestampFormatString\[dq]: \[dq]yyyy-MM-dd_HH\[rs]\[rs]:mm\[rs]\[rs]:ss\[dq],
        \[dq]FrequentSuffix\[dq]: \[dq]frequently\[dq],
        \[dq]HourlySuffix\[dq]: \[dq]hourly\[dq],
        \[dq]DailySuffix\[dq]: \[dq]daily\[dq],
        \[dq]WeeklySuffix\[dq]: \[dq]weekly\[dq],
        \[dq]MonthlySuffix\[dq]: \[dq]monthly\[dq],
        \[dq]YearlySuffix\[dq]: \[dq]yearly\[dq]
      },
      \[dq]SnapshotTiming\[dq]: {
        \[dq]UseLocalTime\[dq]: true,
        \[dq]FrequentPeriod\[dq]: 15,
        \[dq]HourlyMinute\[dq]: 0,
        \[dq]DailyTime\[dq]: \[dq]00:00:00\[dq],
        \[dq]WeeklyDay\[dq]: 1,
        \[dq]WeeklyTime\[dq]: \[dq]00:00:00\[dq],
        \[dq]MonthlyDay\[dq]: 1,
        \[dq]MonthlyTime\[dq]: \[dq]00:00:00\[dq],
        \[dq]YearlyMonth\[dq]: 1,
        \[dq]YearlyDay\[dq]: 1,
        \[dq]YearlyTime\[dq]: \[dq]00:00:00\[dq]
      }
    }
  }
}
\f[R]
.fi
.SS Default Local Configuration (/etc/sanoid/Sanoid.local.json)
.PP
This is the default local configuration shipped with Sanoid.net.
This is the configuration that is intended to be modified by the user
and supercedes the base configuration as described at the top of this
help document.
Keys and values are case-sensitive.
Whitespace not contained within quotation marks is ignored.
.PP
This configuration also defaults to NO actions being taken, so that a
fresh install does not result in changes being made to ZFS.
See the sections above for explanations of what each setting does.
.PP
This configuration includes a template called \[lq]production,\[rq]
which has the same settings as the default template, on a new install.
Additional templates can be added and removed, and the
\[lq]production\[rq] template can be modified or removed as desired by
the user.
.PP
It is strongly recommended that all modifications to this configuration
be made using the configuration console, by invoking Sanoid.net with the
--config-console command-line argument.
This will ensure that your settings are valid and that in-use templates
are not accidentally removed.
.PP
Note that, at this time, the monitoring functionality is not yet
implemented, so the Monitoring section will be ignored.
.IP
.nf
\f[C]
{
  \[dq]$schema\[dq]: \[dq]Sanoid.local.schema.json\[dq],
  \[dq]$id\[dq]: \[dq]Sanoid.local.json\[dq],
  \[dq]$comments\[dq]: \[dq]Values specified here supersede and extend the base configuration in Sanoid.json.\[dq],
  \[dq]CacheDirectory\[dq]: \[dq]var/cache/sanoid\[dq],
  \[dq]ZfsPath\[dq]: \[dq]/usr/local/sbin/zfs\[dq], 
  \[dq]ZpoolPath\[dq]: \[dq]/usr/local/sbin/zpool\[dq], 
  \[dq]TakeSnapshots\[dq]: false,
  \[dq]PruneSnapshots\[dq]: false,
  \[dq]DryRun\[dq]: false,
  \[dq]Monitoring\[dq]: {
    \[dq]Nagios\[dq]: {
      \[dq]$comments\[dq]: \[dq]Nagios-specific monitoring options\[dq],
      \[dq]MonitorType\[dq]: \[dq]Nagios\[dq],
      \[dq]Capacity\[dq]: false,
      \[dq]Health\[dq]: false,
      \[dq]Snapshots\[dq]: false
    }
  },
  \[dq]Templates\[dq]: {
    \[dq]production\[dq]: {
      \[dq]Formatting\[dq]: {
        \[dq]ComponentSeparator\[dq]: \[dq]_\[dq],
        \[dq]Prefix\[dq]: \[dq]autosnap\[dq],
        \[dq]TimestampFormatString\[dq]: \[dq]yyyy-MM-dd_HH\[rs]\[rs]:mm\[rs]\[rs]:ss\[dq],
        \[dq]FrequentSuffix\[dq]: \[dq]frequently\[dq],
        \[dq]HourlySuffix\[dq]: \[dq]hourly\[dq],
        \[dq]DailySuffix\[dq]: \[dq]daily\[dq],
        \[dq]WeeklySuffix\[dq]: \[dq]weekly\[dq],
        \[dq]MonthlySuffix\[dq]: \[dq]monthly\[dq],
        \[dq]YearlySuffix\[dq]: \[dq]yearly\[dq]
      },
      \[dq]SnapshotTiming\[dq]: {
        \[dq]UseLocalTime\[dq]: true,
        \[dq]FrequentPeriod\[dq]: 15,
        \[dq]HourlyMinute\[dq]: 0,
        \[dq]DailyTime\[dq]: \[dq]00:00:00\[dq],
        \[dq]WeeklyDay\[dq]: 1,
        \[dq]WeeklyTime\[dq]: \[dq]00:00:00\[dq],
        \[dq]MonthlyDay\[dq]: 1,
        \[dq]MonthlyTime\[dq]: \[dq]00:00:00\[dq],
        \[dq]YearlyMonth\[dq]: 1,
        \[dq]YearlyDay\[dq]: 1,
        \[dq]YearlyTime\[dq]: \[dq]00:00:00\[dq]
      }
    }
  }
}
\f[R]
.fi
.SH AUTHORS
Brandon Thetford.


Sanoid/Sanoid.cs
// LICENSE:
// 
// This software is licensed for use under the Free Software Foundation's GPL v3.0 license, as retrieved
// from http://www.gnu.org/licenses/gpl-3.0.html on 2014-11-17.  A copy should also be available in this
// project's Git repository at https://github.com/jimsalterjrs/sanoid/blob/master/LICENSE.

using System.Collections.Concurrent;
using System.Text.Json;
using Microsoft.Extensions.Configuration;
using PowerArgs;
using Sanoid.Common.Logging;
using Sanoid.Interop.Concurrency;
using Sanoid.Interop.Libc;
using Sanoid.Interop.Libc.Enums;
using Sanoid.Interop.Zfs.ZfsCommandRunner;
using Sanoid.Interop.Zfs.ZfsTypes;
using Sanoid.Settings.Settings;

namespace SnapsInAZfs;

[UsedImplicitly]
internal class Program
{
    // Note that logging will be at whatever level is defined in Sanoid.nlog.json until configuration is initialized, regardless of command-line parameters.
    // Desired logging parameters should be set in Sanoid.nlog.json
    private static readonly Logger Logger = LogManager.GetCurrentClassLogger( );
    internal static SanoidSettings? Settings;

    public static async Task<int> Main( string[] argv )
    {
        LoggingSettings.ConfigureLogger( );

        DateTimeOffset currentTimestamp = DateTimeOffset.Now;

        Logger.Trace( "Parsing command-line arguments" );
        CommandLineArguments? args = Args.Parse<CommandLineArguments>( argv );

        // The nullability context in PowerArgs is wrong, so this absolutely can be null
        // ReSharper disable once ConditionIsAlwaysTrueOrFalseAccordingToNullableAPIContract
        if ( args is null )
        {
            Logger.Trace( "Help argument provided. Exiting." );
            return (int)Errno.ECANCELED;
        }

        if ( args.Help )
        {
            Logger.Trace( "Help argument provided. Exiting." );
            return (int)Errno.ECANCELED;
        }

        if ( args.Version )
        {
            Logger.Trace( "Version argument provided. Exiting." );
            return (int)Errno.ECANCELED;
        }

        SetCommandLineLoggingOverride( args );

        using Mutexes mutexes = Mutexes.Instance;
        using MutexAcquisitionResult mutexResult = Mutexes.GetAndWaitMutex( "Global\\Sanoid.net" );

        switch ( mutexResult.ErrorCode )
        {
            case MutexAcquisitionErrno.Success:
                Logger.Debug( "Succesfully acquired global mutex." );
                break;
            case MutexAcquisitionErrno.IoException:
                Logger.Fatal( mutexResult.Exception, "Exiting due to IOException: {0}", mutexResult.Exception.Message );
                return (int)mutexResult.ErrorCode;
            case MutexAcquisitionErrno.AbandonedMutex:
                Logger.Fatal( mutexResult.Exception, "A previous instance of Sanoid.net exited without properly releasing the mutex. Sanoid.net will now exit after releasing the abandoned mutex. Try running again." );
                return (int)mutexResult.ErrorCode;
            case MutexAcquisitionErrno.WaitHandleCannotBeOpened:
                Logger.Fatal( mutexResult.Exception, "Unable to acquire mutex. Sanoid.net will exit." );
                return (int)mutexResult.ErrorCode;
            case MutexAcquisitionErrno.PossiblyNullMutex:
                Logger.Fatal( "Unable to acquire mutex. Sanoid.net will exit." );
                return (int)mutexResult.ErrorCode;
            case MutexAcquisitionErrno.AnotherProcessIsBusy:
                Logger.Fatal( "Another Sanoid.net process is running. This process will terminate." );
                return (int)mutexResult.ErrorCode;
            case MutexAcquisitionErrno.InvalidMutexNameRequested:
                return (int)mutexResult.ErrorCode;
            default:
                Logger.Fatal( mutexResult.Exception, "Failed to get mutex. Exiting." );
                return (int)mutexResult.ErrorCode;
        }

        // Configuration is built in the following order from various sources.
        // Configurations from all sources are merged, and the final configuration that will be used is the result of the merged configurations.
        // If conflicting items exist in multiple configuration sources, the configuration of the configuration source added latest will
        // override earlier values.
        // Note that nlog-specific configuration is separate, in Sanoid.nlog.json, and is not affected by the configuration specified below,
        // and is loaded/parsed FIRST, before any configuration specified below.
        // See the Sanoid.Common.Logging.LoggingSettings class for nlog configuration details.
        // See documentation for a more detailed explanation with examples.
        // Configuration order:
        // 1. /usr/local/share/Sanoid.net/Sanoid.json   #(Required - Base configuration - Should not be modified by the user)
        // 2. /etc/sanoid/Sanoid.local.json
        // 3. ~/.config/Sanoid.net/Sanoid.user.json     #(Located in executing user's home directory)
        // 4. ./Sanoid.local.json                       #(Located in Sanoid.net's working directory)
        // 6. Command-line arguments passed on invocation of Sanoid.net
        Logger.Debug( "Getting base configuration from files" );
        IConfigurationRoot rootConfiguration = new ConfigurationBuilder( )
                                           #if WINDOWS
                                               .AddJsonFile( "Sanoid.json", true, false )
                                               .AddJsonFile( "Sanoid.local.json", true, false )
                                           #else
                                               .AddJsonFile( "/usr/local/share/Sanoid.net/Sanoid.json", true, false )
                                               .AddJsonFile( "/etc/sanoid/Sanoid.local.json", true, false )
                                               .AddJsonFile( Path.Combine( Path.GetFullPath( Environment.GetEnvironmentVariable( "HOME" ) ?? "~/" ), ".config/Sanoid.net/Sanoid.user.json" ), true, false )
                                           #endif
                                           #if ALLOW_ADJACENT_CONFIG_FILE
                                               .AddJsonFile( "Sanoid.local.json", true, false )
                                           #endif
                                               .Build( );

        Logger.Trace( "Building settings objects from IConfiguration" );
        try
        {
            Settings = rootConfiguration.Get<SanoidSettings>( ) ?? throw new InvalidOperationException( );
        }
        catch ( Exception ex )
        {
            Logger.Fatal( ex, "Unable to parse settings from JSON" );
            return (int)Errno.EFTYPE;
        }

        ApplyCommandLineArgumentOverrides( in args );

        Logger.Trace( "Getting ZFS command runner for the current environment" );
        IZfsCommandRunner zfsCommandRunner = Environment.OSVersion.Platform switch
        {
            PlatformID.Unix => new ZfsCommandRunner( Settings.ZfsPath, Settings.ZpoolPath ),
            _ => new DummyZfsCommandRunner( )
        };

        Logger.Debug( "Using Settings: {0}", JsonSerializer.Serialize( Settings ) );

        ZfsTasks.CheckZfsPropertiesSchemaResult schemaCheckResult = await ZfsTasks.CheckZfsPoolRootPropertiesSchemaAsync( zfsCommandRunner, args ).ConfigureAwait( true );

        Logger.Debug( "Result of schema check is: {0}", JsonSerializer.Serialize( schemaCheckResult ) );

        // Check
        switch ( args )
        {
            case { CheckZfsProperties: true } when !schemaCheckResult.MissingPropertiesFound:
            {
                // Requested check and no properties were missing.
                // Return 0
                return (int)Errno.EOK;
            }
            case { CheckZfsProperties: true } when schemaCheckResult.MissingPropertiesFound:
            {
                // Requested check and some properties were missing.
                // Return ENOATTR (1093)
                return (int)Errno.ENOATTR;
            }
            case { CheckZfsProperties: false, PrepareZfsProperties: false } when schemaCheckResult.MissingPropertiesFound:
            {
                // Did not request check or update (normal run) but properties were missing.
                // Cannot safely do anything useful
                // Log a fatal error and exit with ENOATTR
                Logger.Fatal( "Missing properties were found in zfs. Cannot continue. Exiting" );
                return (int)Errno.ENOATTR;
            }
            case { PrepareZfsProperties: true }:
            {
                // Requested schema update
                // Run the update and return EOK or ENOATTR based on success of the updates
                return ZfsTasks.UpdateZfsDatasetSchema( Settings.DryRun, schemaCheckResult.PoolRootsWithPropertyValidities, zfsCommandRunner )
                    ? (int)Errno.EOK
                    : (int)Errno.ENOATTR;
            }
        }

        if ( args.ConfigConsole )
        {
            ConfigConsole.ConfigConsole.RunConsoleInterface( zfsCommandRunner );
            return (int)Errno.EOK;
        }

        ConcurrentDictionary<string, Dataset> datasets = new( );
        ConcurrentDictionary<string, Snapshot> snapshots = new( );

        Logger.Debug( "Getting remaining datasets and all snapshots from ZFS" );

        await ZfsTasks.GetDatasetsAndSnapshotsFromZfsAsync( zfsCommandRunner, datasets, snapshots ).ConfigureAwait( true );

        Logger.Debug( "Finished getting datasets and snapshots from ZFS" );

        // Handle taking new snapshots, if requested
        if ( Settings is { TakeSnapshots: true } )
        {
            Logger.Debug( "TakeSnapshots is true. Taking configured snapshots using timestamp {0:O}", currentTimestamp );
            ZfsTasks.TakeAllConfiguredSnapshots( zfsCommandRunner, Settings, currentTimestamp, datasets );
        }
        else
        {
            Logger.Info( "Not taking snapshots" );
        }

        // Handle pruning old snapshots, if requested
        if ( Settings is { PruneSnapshots: true } )
        {
            Logger.Debug( "PruneSnapshots is true. Pruning configured snapshots" );
            await ZfsTasks.PruneAllConfiguredSnapshotsAsync( zfsCommandRunner, Settings, datasets ).ConfigureAwait( true );
        }
        else
        {
            Logger.Info( "Not pruning snapshots" );
        }

        Mutexes.DisposeMutexes( );

        return (int)Errno.EOK;
    }

    /// <summary>
    ///     Overrides configuration values specified in configuration files or environment variables with arguments supplied on
    ///     the CLI
    /// </summary>
    /// <param name="args"></param>
    public static void ApplyCommandLineArgumentOverrides( in CommandLineArguments args )
    {
        Logger.Debug( "Overriding settings using arguments from command line." );

        // Let's go through all args in an order that makes sense
        if ( !string.IsNullOrEmpty( args.CacheDir ) )
        {
            Logger.Debug( "CacheDir argument specified. Value: {0}", args.CacheDir );
            string canonicalCacheDirPath = NativeMethods.CanonicalizeFileName( args.CacheDir );
            Logger.Trace( "CacheDir canonical path: {0}", canonicalCacheDirPath );
            if ( !Directory.Exists( canonicalCacheDirPath ) )
            {
                string badDirectoryMessage = $"CacheDir argument value {canonicalCacheDirPath} is a non-existent directory. Program will terminate.";
                Logger.Error( badDirectoryMessage );
                throw new DirectoryNotFoundException( badDirectoryMessage );
            }

            if ( NativeMethods.EuidAccess( canonicalCacheDirPath, UnixFileTestMode.Read ) != 0 )
            {
                string cantReadDirMessage = $"CacheDir {canonicalCacheDirPath} is not readable by the current user {Environment.UserName}. Program will terminate.";
                Logger.Error( cantReadDirMessage );
                throw new UnauthorizedAccessException( cantReadDirMessage );
            }

            if ( NativeMethods.EuidAccess( canonicalCacheDirPath, UnixFileTestMode.Write ) != 0 )
            {
                string cantWriteDirMessage = $"CacheDir {canonicalCacheDirPath} is not writeable by the current user {Environment.UserName}. Program will terminate.";
                Logger.Error( cantWriteDirMessage );
                throw new UnauthorizedAccessException( cantWriteDirMessage );
            }

            if ( Settings is not null )
            {
                Settings.CacheDirectory = args.CacheDir;
            }

            Logger.Trace( "CacheDirectory is now {0}", canonicalCacheDirPath );
        }

        Settings!.DryRun |= args.DryRun;
        Settings.TakeSnapshots = ( Settings.TakeSnapshots | args.TakeSnapshots ) & !args.NoTakeSnapshots;
        Settings.PruneSnapshots = ( Settings.PruneSnapshots | args.PruneSnapshots ) & !args.NoPruneSnapshots;
    }

    private static void SetCommandLineLoggingOverride( CommandLineArguments args )
    {
        if ( args.Debug )
        {
            LoggingSettings.OverrideConsoleLoggingLevel( LogLevel.Debug );
        }

        if ( args.Quiet )
        {
            LoggingSettings.OverrideConsoleLoggingLevel( LogLevel.Warn );
        }

        if ( args.ReallyQuiet )
        {
            LoggingSettings.OverrideConsoleLoggingLevel( LogLevel.Off );
        }

        if ( args.Trace )
        {
            LoggingSettings.OverrideConsoleLoggingLevel( LogLevel.Trace );
        }
    }
}
